File: tools/fpgainfo/sysfs.py
# Copyright(c) 2017, Intel Corporation
##
# Redistribution  and  use  in source  and  binary  forms,  with  or  without
# modification, are permitted provided that the following conditions are met:
##
# * Redistributions of  source code  must retain the  above copyright notice,
# this list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
# this list of conditions and the following disclaimer in the documentation
# and/or other materials provided with the distribution.
# * Neither the name  of Intel Corporation  nor the names of its contributors
# may be used to  endorse or promote  products derived  from this  software
# without specific prior written permission.
##
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
# IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
# LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
# CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
# SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
# INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
# CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.
import glob
import json
import logging
import os
import re
import uuid

pattern = ('.*/\d+:(?P<bus>\w{2}):'
           '(?P<device>\d{2})\.(?P<function>\d).*')
bdf_pattern = re.compile(pattern)

ROOT_PATH = '/sys/class/fpga'
FPGA_DEVICE = os.path.join(ROOT_PATH, 'intel-fpga-dev.{instance_id}')
FME_DEVICE = os.path.join(FPGA_DEVICE, 'intel-fpga-fme.{instance_id}')
PORT_DEVICE = os.path.join(FPGA_DEVICE, 'intel-fpga-port.{instance_id}')


def read_bdf(path):
    symlink = os.readlink(path)
    m = bdf_pattern.match(symlink)
    data = m.groupdict() if m else {}
    return dict([(k, int(v, 16)) for (k, v) in data.iteritems()])


class sysfs_filter(object):
    def __init__(self, **kwargs):
        self._bus = None
        self._device = None
        self._function = None
        for attr in ('bus', 'device', 'function'):
            value = kwargs.get(attr)
            if isinstance(value, str) or isinstance(value, unicode):
                value = int(value, 16)
            setattr(self, '_{}'.format(attr), value)

    def __call__(self, resource):
        if self._bus is not None and self._bus != resource.bus:
            return False
        if self._device is not None and self._device != resource.device:
            return False
        if self._function is not None and self._function != resource.function:
            return False
        return True


def add_static_property(sysfs_path):
    return property(lambda s_: s_.parse_sysfs(sysfs_path),
                    lambda s_, v_: s_.write_sysfs(v_, sysfs_path))


class sysfs_resource(object):
    def __init__(self, path, instance_id, **kwargs):
        self._path = path
        self._instance_id = instance_id
        self._bus = kwargs.get('bus')
        self._device = kwargs.get('device')
        self._function = kwargs.get('function')

    def to_dict(self, include_bdf=True):
        data = {}
        for k, v in vars(self).iteritems():
            if hasattr(v.__class__, 'to_dict'):
                data[k] = v.to_dict(False)
        for k, v in vars(self.__class__).iteritems():
            if isinstance(v, property):
                data[k] = v.fget(self)
        for k, v in vars(self).iteritems():
            if isinstance(v, property):
                data[k] = v.fget(self)
        if include_bdf:
            root_data = {}
            root_data["class_path"] = self.sysfs_path
            root_data["dev_path"] = os.path.realpath(self.sysfs_path)
            root_data["bus"] = self.bus
            root_data["device"] = self.device
            root_data["function"] = self.function
            data["pcie_info"] = root_data

        return data

    def to_json(self):
        return json.dumps(self.to_dict(), indent=4, sort_keys=False)

    def sysfs_path_exists(self, *paths):
        return os.path.exists(os.path.join(self._path, *paths))

    def read_sysfs(self, *paths):
        filepath = os.path.join(self._path, *paths)
        if not os.path.exists(filepath):
            print("WARNING: {} not found".format(filepath))
            return None
        with open(filepath, 'r') as fd:
            return fd.read().strip()

    def write_sysfs(self, value, *paths):
        filepath = os.path.join(self._path, *paths)
        if not os.path.exists(filepath):
            print("WARNING: {} not found".format(filepath))
            return None
        with open(filepath, 'w') as fd:
            return fd.write(value)

    def parse_sysfs(self, *paths):
        value = self.read_sysfs(*paths)
        try:
            value = int(value)
        except ValueError:
            try:
                value = int(value, 16)
            except Exception:
                logging.warn("Could not parse value: {}".format(value))
        finally:
            return value

    def print_info(self, label, *props, **kwargs):
        print(label.upper())
        print('{:22} : {}'.format('Class Path', self.sysfs_path))
        print(
            '{:22} : {}'.format(
                'Device Path',
                os.path.realpath(
                    self.sysfs_path)))
        print('{:22} : 0x{:02X}'.format('Bus', self.bus))
        print('{:22} : 0x{:02X}'.format('Device', self.device))
        print('{:22} : 0x{:02X}'.format('Function', self.function))
        for prop in filter(
                lambda p: hasattr(
                    self.__class__, p) or hasattr(self, p),
                props):
            prop_value = getattr(self, prop)
            prop_value = kwargs.get(prop, lambda p: p)(prop_value)
            print(
                u'{:22} : {}'.format(
                    ' '.join([p.capitalize() for p in prop.split('_')]),
                    prop_value))
        print('\n')

    @property
    def sysfs_path(self):
        return self._path

    @property
    def instance_id(self):
        return self._instance_id

    @property
    def bus(self):
        return self._bus

    @property
    def device(self):
        return self._device

    @property
    def function(self):
        return self._function


class pr_feature(sysfs_resource):
    @property
    def interface_id(self):
        return str(uuid.UUID(self.read_sysfs("interface_id")))


class power_mgmt_feature(sysfs_resource):
    consumed = add_static_property("consumed")


class thermal_feature(sysfs_resource):
    temperature = add_static_property("temperature")
    threshold1 = add_static_property("threshold1")
    threshold1_policy = add_static_property("threshold1_policy")
    threshold1_reached = add_static_property("threshold1_reached")
    threshold2 = add_static_property("threshold2")
    threshold2_reached = add_static_property("threshold2_reached")
    threshold_trip = add_static_property("threshold_trip")


class errors_feature(sysfs_resource):
    error_files = {}
    revision = add_static_property("revision")

    def __init__(self, path, instance_id, **kwargs):
        super(errors_feature, self).__init__(path, instance_id, **kwargs)
        self.name = None
        self.errors_file = None
        self.clear_file = None
        self.error_classes = []
        self.valid_error_files = []
        # first check the error files for the given revision
        errfiles = self.error_files[self.revision]
        if all([self.sysfs_path_exists(f) for f in errfiles]):
            self.add_error_props(*errfiles)
        else:
            for rev, errfiles in self.error_files.iteritems():
                self.add_error_props(*errfiles)

    def print_info(self, label, *props, **kwargs):
        super(errors_feature, self).print_info(label, *self.valid_error_files)

    def add_error_props(self, *errfiles):
        for err_file in errfiles:
            base_name = os.path.basename(err_file)
            if self.sysfs_path_exists(
                    err_file) and not hasattr(self, base_name):
                self.valid_error_files.append(base_name)
                setattr(self,
                        os.path.basename(err_file),
                        property(lambda self_: self_.parse_sysfs(err_file)))

    def clear(self):
        value = self.parse_sysfs(self.errors_file)
        try:
            if value:
                self.write_sysfs(hex(value), self.clear_file)
            return True
        except IOError:
            logging.warn(
                "Could not clear errors: {}."
                "Are you running as root?".format(
                    self.clear_file))
        return False


class fme_errors(errors_feature):
    error_files = {
        0: ["fme-errors/errors",
            "fme-errors/first_error",
            "fme-errors/next_error",
            "bbs_errors",
            "gbs_errors",
            "pcie0_errors",
            "pcie1_errors"],
        1: ["fme-errors/errors",
            "fme-errors/first_error",
            "fme-errors/next_error",
            "catfatal_errors",
            "nonfatal_errors",
            "pcie0_errors",
            "pcie1_errors"]}

    def __init__(self, path, instance_id, **kwargs):
        super(fme_errors, self).__init__(path, instance_id, **kwargs)
        self.name = "fme errors"
        self.errors_file = "fme-errors/errors"
        self.clear_file = "fme-errors/clear"


class port_errors(errors_feature):
    error_files = {
        0: ["errors",
            "first_error"],
        1: ["errors",
            "first_error"]}

    def __init__(self, path, instance_id, **kwargs):
        super(port_errors, self).__init__(path, instance_id, **kwargs)
        self.name = "port errors"
        self.errors_file = "errors"
        self.clear_file = "clear"


class fme_info(sysfs_resource):
    def __init__(self, path, instance_id, **kwargs):
        super(fme_info, self).__init__(path, instance_id, **kwargs)
        self.pr = pr_feature(os.path.join(path, 'pr'), instance_id, **kwargs)
        self.power_mgmt = power_mgmt_feature(os.path.join(path, 'power_mgmt'),
                                             instance_id,
                                             **kwargs)
        self.thermal_mgmt = thermal_feature(os.path.join(path, "thermal_mgmt"),
                                            instance_id,
                                            **kwargs)
        self.errors = fme_errors(os.path.join(path, "errors"),
                                 instance_id,
                                 **kwargs)

    @property
    def version(self):
        return self.read_sysfs("version")

    @property
    def ports_num(self):
        return self.read_sysfs("ports_num")

    @property
    def socket_id(self):
        return self.read_sysfs("socket_id")

    @property
    def bitstream_id(self):
        return self.read_sysfs("bitstream_id")

    @property
    def bitstream_metadata(self):
        return self.read_sysfs("bitstream_metadata")

    @property
    def object_ID(self):
        value = self.read_sysfs("dev")
        valueList = value.split(":")
        major = valueList[0]
        minor = valueList[1]
        objID = ((int(major) & 0xFFF) << 20) | (int(minor) & 0xFFFFF)
        return hex(objID) + "   FPGA_DEVICE"


class port_info(sysfs_resource):
    def __init__(self, path, instance_id, **kwargs):
        super(port_info, self).__init__(path, instance_id, **kwargs)
        self.errors = port_errors(os.path.join(path, "errors"),
                                  instance_id,
                                  **kwargs)

    @property
    def afu_id(self):
        return str(uuid.UUID(self.read_sysfs("afu_id")))

    @property
    def object_ID(self):
        value = self.read_sysfs("dev")
        valueList = value.split(":")
        major = valueList[0]
        minor = valueList[1]
        objID = ((int(major) & 0xFFF) << 20) | (int(minor) & 0xFFFFF)
        return hex(objID) + "   FPGA_ACCELERATOR"


class sysfsinfo(object):
    def __init__(self):
        self._fmelist = []
        self._portlist = []
        sysfs_paths = glob.glob(
            FPGA_DEVICE.format(instance_id='*'))
        if not sysfs_paths:
            print "WARNING: No sysfs paths found"
        for path in sysfs_paths:
            bdf = read_bdf(os.path.join(path, 'device'))
            # strip {instance_id} from the template FPGA_DEVICE
            # socket id is what comes after this in the real path
            instance_id = path.strip(FPGA_DEVICE.strip('{instance_id}'))
            sysfs_fme = FME_DEVICE.format(instance_id=instance_id)
            sysfs_port = PORT_DEVICE.format(instance_id=instance_id)
            self._fmelist.append(fme_info(sysfs_fme, instance_id, **bdf))
            self._portlist.append(port_info(sysfs_port, instance_id, **bdf))

    def fme(self, **kwargs):
        return filter(sysfs_filter(**kwargs), self._fmelist)

    def port(self, **kwargs):
        return filter(sysfs_filter(**kwargs), self._portlist)


if __name__ == "__main__":
    import argparse
    import pprint
    parser = argparse.ArgumentParser()
    parser.add_argument('-b', '--bus')
    parser.add_argument('-d', '--device')
    parser.add_argument('-f', '--function')

    args = parser.parse_args()
    info = sysfsinfo()
    fmelist = info.fme(**vars(args))
    portlist = info.port(**vars(args))
    pprint.pprint([f.to_dict() for f in fmelist])

File: libopae/src/sysfs.c
// Copyright(c) 2017, Intel Corporation
//
// Redistribution  and  use  in source  and  binary  forms,  with  or  without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of  source code  must retain the  above copyright notice,
//   this list of conditions and the following disclaimer.
// * Redistributions in binary form must reproduce the above copyright notice,
//   this list of conditions and the following disclaimer in the documentation
//   and/or other materials provided with the distribution.
// * Neither the name  of Intel Corporation  nor the names of its contributors
//   may be used to  endorse or promote  products derived  from this  software
//   without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING,  BUT NOT LIMITED TO,  THE
// IMPLIED WARRANTIES OF  MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED.  IN NO EVENT  SHALL THE COPYRIGHT OWNER  OR CONTRIBUTORS BE
// LIABLE  FOR  ANY  DIRECT,  INDIRECT,  INCIDENTAL,  SPECIAL,  EXEMPLARY,  OR
// CONSEQUENTIAL  DAMAGES  (INCLUDING,  BUT  NOT LIMITED  TO,  PROCUREMENT  OF
// SUBSTITUTE GOODS OR SERVICES;  LOSS OF USE,  DATA, OR PROFITS;  OR BUSINESS
// INTERRUPTION)  HOWEVER CAUSED  AND ON ANY THEORY  OF LIABILITY,  WHETHER IN
// CONTRACT,  STRICT LIABILITY,  OR TORT  (INCLUDING NEGLIGENCE  OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE,  EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif // HAVE_CONFIG_H

#include <pthread.h>
#include <unistd.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include <fcntl.h>
#include <errno.h>
#include <opae/types.h>

#include "safe_string/safe_string.h"

#include "types_int.h"
#include "sysfs_int.h"
#include "log_int.h"
#include "common_int.h"

//
// sysfs access (read/write) functions
//

fpga_result sysfs_read_int(const char *path, int *i)
{
	int fd;
	int res;
	char buf[SYSFS_PATH_MAX];
	int b;

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed", path);
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek failed");
		goto out_close;
	}

	b = 0;

	do {
		res = read(fd, buf+b, sizeof(buf)-b);
		if (res <= 0) {
			FPGA_MSG("Read from %s failed", path);
			goto out_close;
		}
		b += res;
		if (((unsigned)b > sizeof(buf)) || (b <= 0)) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}
	} while (buf[b-1] != '\n' && buf[b-1] != '\0' && (unsigned)b < sizeof(buf));

	// erase \n
	buf[b-1] = 0;

	*i = atoi(buf);

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

fpga_result sysfs_read_u32(const char *path, uint32_t *u)
{
	int fd;
	int res;
	char buf[SYSFS_PATH_MAX];
	int b;

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed", path);
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek failed");
		goto out_close;
	}

	b = 0;

	do {
		res = read(fd, buf+b, sizeof(buf)-b);
		if (res <= 0) {
			FPGA_MSG("Read from %s failed", path);
			goto out_close;
		}
		b += res;
		if (((unsigned)b > sizeof(buf)) || (b <= 0)) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}
	} while (buf[b-1] != '\n' && buf[b-1] != '\0' && (unsigned)b < sizeof(buf));

	// erase \n
	buf[b-1] = 0;

	*u = strtoul(buf, NULL, 0);

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

// read tuple separated by 'sep' character
fpga_result sysfs_read_u32_pair(const char *path, uint32_t *u1, uint32_t *u2,
				 char sep)
{
	int fd;
	int res;
	char buf[SYSFS_PATH_MAX];
	int b;
	char *c;
	uint32_t x1, x2;

	if (sep == '\0') {
		FPGA_MSG("invalid separation character");
		return FPGA_INVALID_PARAM;
	}

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed", path);
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek failed");
		goto out_close;
	}

	b = 0;

	do {
		res = read(fd, buf+b, sizeof(buf)-b);
		if (res <= 0) {
			FPGA_MSG("Read from %s failed", path);
			goto out_close;
		}
		b += res;
		if (((unsigned)b > sizeof(buf)) || (b <= 0)) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}
	} while (buf[b-1] != '\n' && buf[b-1] != '\0' && (unsigned)b < sizeof(buf));

	// erase \n
	buf[b-1] = 0;

	// read first value
	x1 = strtoul(buf, &c, 0);
	if (*c != sep) {
		FPGA_MSG("couldn't find separation character '%c' in '%s'", sep, path);
		goto out_close;
	}
	// read second value
	x2 = strtoul(c+1, &c, 0);
	if (*c != '\0') {
		FPGA_MSG("unexpected character '%c' in '%s'", *c, path);
		goto out_close;
	}

	*u1 = x1;
	*u2 = x2;

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

fpga_result __FIXME_MAKE_VISIBLE__ sysfs_read_u64(const char *path, uint64_t *u)
{
	int fd                     = -1;
	int res                    = 0;
	char buf[SYSFS_PATH_MAX]   = {0};
	int b                      = 0;

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed", path);
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek failed");
		goto out_close;
	}

	do {
		res = read(fd, buf+b, sizeof(buf)-b);
		if (res <= 0) {
			FPGA_MSG("Read from %s failed", path);
			goto out_close;
		}
		b += res;
		if (((unsigned)b > sizeof(buf)) || (b <= 0)) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}
	} while (buf[b-1] != '\n' && buf[b-1] != '\0' && (unsigned)b < sizeof(buf));

	// erase \n
	buf[b-1] = 0;

	*u = strtoull(buf, NULL, 0);

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

fpga_result __FIXME_MAKE_VISIBLE__ sysfs_write_u64(const char *path, uint64_t u)
{
	int fd                     = -1;
	int res                    = 0;
	char buf[SYSFS_PATH_MAX]   = {0};
	int b                      = 0;

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_WRONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed: %s", path, strerror(errno));
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek: %s", strerror(errno));
		goto out_close;
	}

	snprintf_s_l(buf, sizeof(buf), "0x%lx", u);

	do {
		res = write(fd, buf + b, sizeof(buf) -b);
		if (res <= 0) {
			FPGA_ERR("Failed to write");
			goto out_close;
		}
		b += res;

		if ((unsigned)b > sizeof(buf) || b <= 0) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}

	} while (buf[b - 1] != '\n' && buf[b - 1] != '\0' && (unsigned)b < sizeof(buf));

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

fpga_result sysfs_read_guid(const char *path, fpga_guid guid)
{
	int fd;
	int res;
	char buf[SYSFS_PATH_MAX];
	int b;

	int i;
	char tmp;
	unsigned octet;

	if (path == NULL) {
		FPGA_ERR("Invalid input path");
		return FPGA_INVALID_PARAM;
	}

	fd = open(path, O_RDONLY);
	if (fd < 0) {
		FPGA_MSG("open(%s) failed", path);
		return FPGA_NOT_FOUND;
	}

	if ((off_t)-1 == lseek(fd, 0, SEEK_SET)) {
		FPGA_MSG("seek failed");
		goto out_close;
	}

	b = 0;

	do {
		res = read(fd, buf+b, sizeof(buf)-b);
		if (res <= 0) {
			FPGA_MSG("Read from %s failed", path);
			goto out_close;
		}
		b += res;
		if (((unsigned)b > sizeof(buf)) || (b <= 0)) {
			FPGA_MSG("Unexpected size reading from %s", path);
			goto out_close;
		}
	} while (buf[b-1] != '\n' && buf[b-1] != '\0' && (unsigned)b < sizeof(buf));

	// erase \n
	buf[b-1] = 0;

	for (i = 0 ; i < 32 ; i += 2) {
		tmp = buf[i+2];
		buf[i+2] = 0;

		octet = 0;
		sscanf_s_u(&buf[i], "%x", &octet);
		guid[i/2] = (uint8_t) octet;

		buf[i+2] = tmp;
	}

	close(fd);
	return FPGA_OK;

out_close:
	close(fd);
	return FPGA_NOT_FOUND;
}

//
// sysfs convenience functions to access device components by device number
//

// FIXME: uses same number for device and FME (may not be true in future)
fpga_result sysfs_get_socket_id(int dev, int subdev, uint8_t *socket_id)
{
	fpga_result result;
	char spath[SYSFS_PATH_MAX];
	int i;

	snprintf_s_ii(spath, SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH
		 SYSFS_FME_PATH_FMT "/"
		 FPGA_SYSFS_SOCKET_ID,
		 dev, subdev);

	i = 0;
	result = sysfs_read_int(spath, &i);
	if (FPGA_OK != result)
		return result;

	*socket_id = (uint8_t) i;

	return FPGA_OK;
}

// FIXME: uses same number for device and PORT (may not be true in future)
fpga_result sysfs_get_afu_id(int dev, int subdev, fpga_guid guid)
{
	char spath[SYSFS_PATH_MAX];

	snprintf_s_ii(spath, SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH
		 SYSFS_AFU_PATH_FMT "/"
		 FPGA_SYSFS_AFU_GUID,
		 dev, subdev);

	return sysfs_read_guid(spath, guid);
}

fpga_result sysfs_get_pr_id(int dev, int subdev, fpga_guid guid)
{
	char spath[SYSFS_PATH_MAX];

	snprintf_s_ii(spath, SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH
		 SYSFS_FME_PATH_FMT "/"
		 FPGA_SYSFS_FME_INTERFACE_ID,
		 dev, subdev);

	return sysfs_read_guid(spath, guid);
}

// FIXME: uses same number for device and FME (may not be true in future)
fpga_result sysfs_get_slots(int dev, int subdev, uint32_t *slots)
{
	char spath[SYSFS_PATH_MAX];

	snprintf_s_ii(spath, SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH
		 SYSFS_FME_PATH_FMT "/"
		 FPGA_SYSFS_NUM_SLOTS,
		 dev, subdev);

	return sysfs_read_u32(spath, slots);
}

// FIXME: uses same number for device and FME (may not be true in future)
fpga_result sysfs_get_bitstream_id(int dev, int subdev, uint64_t *id)
{
	char spath[SYSFS_PATH_MAX];

	snprintf_s_ii(spath, SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH
		 SYSFS_FME_PATH_FMT "/"
		 FPGA_SYSFS_BITSTREAM_ID,
		 dev, subdev);

	return sysfs_read_u64(spath, id);
}

// Get port syfs path
fpga_result get_port_sysfs(fpga_handle handle,
				char *sysfs_port)
{

	struct _fpga_token  *_token;
	struct _fpga_handle *_handle  = (struct _fpga_handle *)handle;
	char *p                       = 0;

	if (sysfs_port == NULL) {
		FPGA_ERR("Invalid output pointer");
		return FPGA_INVALID_PARAM;
	}

	if (_handle == NULL) {
		FPGA_ERR("Invalid handle");
		return FPGA_INVALID_PARAM;
	}

	_token = (struct _fpga_token *)_handle->token;
	if (_token == NULL) {
		FPGA_ERR("Token not found");
		return FPGA_INVALID_PARAM;
	}

	p = strstr(_token->sysfspath, FPGA_SYSFS_FME);
	if (NULL == p) {
		FPGA_ERR("Invalid sysfspath in token");
		return FPGA_INVALID_PARAM;
	}

	snprintf_s_ii(sysfs_port, SYSFS_PATH_MAX,
		SYSFS_FPGA_CLASS_PATH SYSFS_AFU_PATH_FMT,
		_token->device_instance, _token->subdev_instance);

	return FPGA_OK;
}

// get fpga device id
fpga_result get_fpga_deviceid(fpga_handle handle,
				uint64_t *deviceid)
{
	struct _fpga_token  *_token      = NULL;
	struct _fpga_handle  *_handle    = (struct _fpga_handle *)handle;
	char sysfs_path[SYSFS_PATH_MAX]  = {0};
	char *p                          = NULL;
	fpga_result result               = FPGA_OK;
	int err                          = 0;

	if (_handle == NULL) {
		FPGA_ERR("Invalid handle");
		return FPGA_INVALID_PARAM;
	}

	if (deviceid == NULL) {
		FPGA_ERR("Invalid input Parameters");
		return FPGA_INVALID_PARAM;
	}

	if (pthread_mutex_lock(&_handle->lock)) {
		FPGA_MSG("Failed to lock handle mutex");
		return FPGA_EXCEPTION;
	}

	_token = (struct _fpga_token *)_handle->token;
	if (_token == NULL) {
		FPGA_ERR("Token not found");
		result = FPGA_INVALID_PARAM;
		goto out_unlock;
	}

	p = strstr(_token->sysfspath, FPGA_SYSFS_FME);
	if (p == NULL) {
		FPGA_ERR("Failed to read sysfs path");
		result = FPGA_NOT_SUPPORTED;
		goto out_unlock;
	}

	snprintf_s_is(sysfs_path,
		 SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH SYSFS_FPGA_FMT "/%s",
		 _token->device_instance,
		 FPGA_SYSFS_DEVICEID);

	result = sysfs_read_u64(sysfs_path, deviceid);
	if (result != 0) {
		FPGA_ERR("Failed to read device ID");
		goto out_unlock;
	}

out_unlock:
	err = pthread_mutex_unlock(&_handle->lock);
	if (err)
		FPGA_ERR("pthread_mutex_unlock() failed: %s", strerror(err));
	return result;
}

// get fpga device id using the sysfs path
fpga_result sysfs_deviceid_from_path(const char *sysfspath,
				uint64_t *deviceid)
{
	char sysfs_path[SYSFS_PATH_MAX]  = {0};
	char *p                          = NULL;
	int device_instance              = 0;
	fpga_result result               = FPGA_OK;

	if (deviceid == NULL) {
		FPGA_ERR("Invalid input Parameters");
		return FPGA_INVALID_PARAM;
	}

	p = strstr(sysfspath, FPGA_SYSFS_FME);
	if (p == NULL) {
		FPGA_ERR("Failed to read sysfs path");
		return FPGA_NOT_SUPPORTED;
	}

	p = strchr(sysfspath, '.');
	if (p == NULL) {
		FPGA_ERR("Failed to read sysfs path");
		return FPGA_NOT_SUPPORTED;
	}

	device_instance = atoi(p + 1);

	snprintf_s_is(sysfs_path,
		 SYSFS_PATH_MAX,
		 SYSFS_FPGA_CLASS_PATH SYSFS_FPGA_FMT "/%s",
		 device_instance,
		 FPGA_SYSFS_DEVICEID);

	result = sysfs_read_u64(sysfs_path, deviceid);
	if (result != 0)
		FPGA_ERR("Failed to read device ID");

	return result;
}

/*
 * The rlpath path is assumed to be of the form:
 * ../../devices/pci0000:5e/0000:5e:00.0/fpga/intel-fpga-dev.0
 */
fpga_result sysfs_sbdf_from_path(const char *sysfspath, int *s, int *b, int *d, int *f)
{
	int res;
	char rlpath[SYSFS_PATH_MAX];
	char *p;

	res = readlink(sysfspath, rlpath, sizeof(rlpath));
	if (-1 == res) {
		FPGA_MSG("Can't read link %s (no driver?)", sysfspath);
		return FPGA_NO_DRIVER;
	}

	// Find the BDF from the link path.
	rlpath[res] = 0;
	p = strrchr(rlpath, '/');
	if (!p) {
		FPGA_MSG("Invalid link %s (no driver?)", rlpath);
		return FPGA_NO_DRIVER;
	}
	*p = 0;
	p = strrchr(rlpath, '/');
	if (!p) {
		FPGA_MSG("Invalid link %s (no driver?)", rlpath);
		return FPGA_NO_DRIVER;
	}
	*p = 0;
	p = strrchr(rlpath, '/');
	if (!p) {
		FPGA_MSG("Invalid link %s (no driver?)", rlpath);
		return FPGA_NO_DRIVER;
	}
	++p;

	//           11
	// 012345678901
	// ssss:bb:dd.f
	*f = (int) strtoul(p+11, NULL, 16);
	*(p + 10) = 0;

	*d = (int) strtoul(p+8, NULL, 16);
	*(p + 7) = 0;

	*b = (int) strtoul(p+5, NULL, 16);
	*(p + 4) = 0;

	*s = (int) strtoul(p, NULL, 16);

	return FPGA_OK;
}

fpga_result sysfs_objectid_from_path(const char *sysfspath, uint64_t *object_id)
{
	char sdevpath[SYSFS_PATH_MAX];
	uint32_t major = 0;
	uint32_t minor = 0;
	fpga_result result;

	snprintf_s_s(sdevpath, SYSFS_PATH_MAX, "%s/dev", sysfspath);

	result = sysfs_read_u32_pair(sdevpath, &major, &minor, ':');
	if (FPGA_OK != result)
		return result;

	*object_id = ((major & 0xFFF) << 20) | (minor & 0xFFFFF);

	return FPGA_OK;
}
